### ISSUE TITLE
Some base64 encoded files in XML request trigger 932140 rule (RCE)

### ISSUE THREAD

**Issue by adrienyhuel** [coreruleset/coreruleset#1785]
### Description

We're using mod security with coreruleset to protect SOAP Webservices.
Some of our webservices have base64 encoded file (mostly pdfs) inline in XML.

Unfortunately, we encountered some base64 encoded pdfs containing value like "xxxx+If+xxxxx==".
This trigger the rule 932140 : Remote Command Execution: Windows FOR/IF Command Found

### Audit Logs / Triggered Rule Numbers

[2020-06-03 11:11:27.460363] [-:error] 1.2.3.4:22982 XteFX1sgrt8ncaRU2zabFQAAAAU [client 1.2.3.4] ModSecurity: Warning. Pattern match "\\\\b(?:if(?:/i)?(?: not)?(?: exist\\\\b| defined\\\\b| errorlevel\\\\b| cmdextversion\\\\b|(?: |\\\\().*(?:\\\\bgeq\\\\b|\\\\bequ\\\\b|\\\\bneq\\\\b|\\\\bleq\\\\b|\\\\bgtr\\\\b|\\\\blss\\\\b|==))|for(?:/[dflr].*)? %+[^ ]+ in\\\\(.*\\\\)\\\\s?do)" at XML. [file "/etc/httpd/modsecurity.d/owasp-modsecurity-crs-3.2.0/rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf"] [line "419"] [id "932140"] [msg "Remote Command Execution: Windows FOR/IF Command Found"] [data "Matched Data: if naw001va/6wfhr/xusqeilz== found within XML:  ? ? 0 123456 example.pdf application/pdf 5911 atraiter professionnels foo foo 2017-04-19t11:29:42 adrien foo foo foo 999999 9en74 if naw001va/6wfhr/xusqeilz== "] [severity "CRITICAL"] [ver "OWASP_CRS/3.2.0"] [tag "application-multi"] [tag "language-shell"] [tag "platform-windows"] [tag "attack-rce"] [tag "OWASP_CRS"] [tag "OWASP_CRS/WEB_ATTACK/COMMAND_INJECTION"] [tag "WASCTC/WASC-31"] [tag "OWASP_TOP_10/A1"] [tag "PCI/6.5.2"] [hostname "busaws-test.mydcpcloud.com"] [uri "/biblos/v201809.0/documentCreation"] [unique_id "XteFX1sgrt8ncaRU2zabFQAAAAU"]

Here, my base64 value in an XML tag was "9en74+If+naW001va/6WFhr/XuSqeiLz==".
It seems that for one reason, mod security transform the plus signs to spaces before applying the regex. In this case, this lead to detect the "if" like an attack.
Without removing the plus signs, the regex won't match.

### Your Environment

* CRS version (e.g., v3.2.0):
* Paranoia level setting: 1
* ModSecurity version (e.g., 2.9.3): mod_security-2.9.3-1.amzn2.0.1.x86_64
* Web Server and version (e.g., apache 2.4.41): apache 2.4.37
* Operating System and version: Amazon Linux 2

### Confirmation

[X] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by adrienyhuel** [comment]
Hello,

Any ETA on this issue ? This leads to false positives in our production system.

**Comment by dune73** [comment]
Hey @adrienyhuel, thank you for reporting and sorry for the lack of feedback. I have added this issue to the agenda of our monthly meeting on Monday. We'll look into it an probably assign it to somebody who will then take care of it.

Do you know how to create a Rule Exclusion as a workaround? I can also help you with that in the meantime.

**Comment by adrienyhuel** [comment]
Hey @dune73, thank you for taking care of this issue.
We currently use modsecurity in Apache with coreruleset, but we did not activated "blocking" mode because of this issue. Fortunately our Apache is behind a WAF, so this mitigate attacks, and we regularly check highest scores in modsecurity logs.

Our CISO would love this rule to not raise high scores and be able to activate blocking mode :smiley:

**Comment by adrienyhuel** [comment]
Hello @dune73 
Did you had time to discuss about that issue in your monthly meeting ?
Thank you

**Comment by dune73** [comment]
Yes, we did. And I volunteered to take it on. But I did not sit down to actually do it so far. It's not quite easy, I dare say.

Here is a possible rule exclusion for this rule on this URI:

```
# ModSec Rule Exclusion: 932140 : Remote Command Execution: Windows FOR/IF Command Found
SecRule REQUEST_URI "@beginsWith /biblos/v201809.0/documentCreation" \ 
     "phase:1,nolog,pass,id:10000,ctl:ruleRemoveById=932140"
```

This should get rid of this FP and allow you to do blocking.

**Comment by dune73** [comment]
OK. This is really tricky.

Our `t:urlDecodeUni` gets rid of the `+` symbols and this leads to this FP. We need to get rid of these symbols or we do not have any hits at all anymore.

Here is the minimal payload to trigger this:

```
curl http://localhost --header "Content-Type: application/xml" --data '<?xml version="1.0" encoding="UTF-8"?><note><pdf>if+foo==</pdf></note>'
```

**Comment by theseion** [comment]
We need to look at this again for two reasons:
1. I don't think we can assume that `==\b` is equivalent (tokenization doesn't require a whitespace because there are usually no keywords that include `=`)
2. The transformation instructions of the rule have changed; the are now `t:none,t:cmdLine` (see https://github.com/SpiderLabs/owasp-modsecurity-crs/pull/590). Because of this the `+` could potentially still be in there and the FP wouldn't trigger in the first place (have not tested this).

**Comment by theseion** [comment]
I just ran the test and, indeed, the removal of `t:urlDecodeUni` fixes the FP for the line you posted above.

I'll add the test against the FP so we'll notice should we make a change to the transformation functions.
